File: javascript/workflow/src/resources/TableStore.ts
/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */
/* eslint-disable @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-assignment */
import { BehaviorSubject } from 'rxjs'

import type { DataTable } from './DataTable.js'
import { Observed } from './Observed.js'

export class TableStore extends Observed {
	private _tables: Map<string, DataTable> = new Map()
	private _tablesObservable = new BehaviorSubject<DataTable[]>([])

	public add(table: DataTable): void {
		let { name } = table
		this._tables.set(name, table)
		table.onChange(() => {
			if (name !== table.name) {
				this._tables.delete(name)
				this._tables.set(table.name, table)
				name = table.name
				this._emitTables()
			}
		})
		this._emitTables()
	}

	public remove(name: string): void {
		this._tables.delete(name)
		this._emitTables()
	}

	public get(name: string): DataTable | undefined {
		return this._tables.get(name)
	}

	public get size(): number {
		return this._tables.size
	}

	public get names(): string[] {
		return [...this._tables.keys()]
	}

	public clear(): void {
		this._tables.clear()
		this._emitTables()
	}

	public get tables(): BehaviorSubject<DataTable[]> {
		return this._tablesObservable
	}

	private _emitTables(): void {
		this._tablesObservable.next([...this._tables.values()])
		this._onChange.next()
	}
}

File: python/datashaper/datashaper/table_store/table_store.py
#
# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project.
#
"""Abstract class to define the table store interface."""
from abc import ABC, abstractmethod

from datashaper.table_store.types import TableContainer


class TableStore(ABC):
    """A table store for storing and retrieving tables."""

    @abstractmethod
    def add(self, name: str, table: TableContainer, tag: str = "default") -> None:
        """Add a table to the store."""

    @abstractmethod
    def get(self, name: str) -> TableContainer:
        """Get a table from the store."""

    @abstractmethod
    def remove(self, name: str) -> None:
        """Remove a table from the store."""

    @abstractmethod
    def list(self, tag: str | None = None) -> list[str]:
        """List the tables in the store."""

    @abstractmethod
    def dispose(self) -> None:
        """Clear the store."""
